src/devices/bus/msx/cart/ram.cpp

```diff
@@ -0,0 +1,320 @@
+		page(0)->install_ram(0x0000, 0x3fff, m_ram.data());
+		page(1)->install_ram(0x4000, 0x7fff, m_ram.data() + 0x4000);
+		page(2)->install_ram(0x8000, 0xbfff, m_ram.data() + 0x8000);
+		page(3)->install_ram(0xc000, 0xffff, m_ram.data() + 0xc000);
+	}
+};
+
+
+class msx_cart_base_mm_ram_device : public msx_cart_base_ram_device
+{
+protected:
+	msx_cart_base_mm_ram_device(const machine_config &mconfig, device_type type, const char *tag, device_t *owner, u32 clock, u32 ram_size)
+		: msx_cart_base_ram_device(mconfig, type, tag, owner, clock, ram_size)
+		, m_nr_banks(ram_size / 0x4000)
+		, m_rambank(*this, "rambank%u", 0U)
+	{ }
+
+	virtual void device_start() override;
+
+	template <int Bank> void bank_w(u8 data);
+
+	u16 m_nr_banks;
+	memory_bank_array_creator<4> m_rambank;
+};
+
+void msx_cart_base_mm_ram_device::device_start()
+{
+	msx_cart_base_ram_device::device_start();
+
+	m_rambank[0]->configure_entries(0, m_nr_banks, m_ram.data(), 0x4000);
+	m_rambank[1]->configure_entries(0, m_nr_banks, m_ram.data(), 0x4000);
+	m_rambank[2]->configure_entries(0, m_nr_banks, m_ram.data(), 0x4000);
+	m_rambank[3]->configure_entries(0, m_nr_banks, m_ram.data(), 0x4000);
+
+	// The MSX system allows multiple devices to react to I/O, so we use taps.
+	io_space().install_write_tap(0xfc, 0xfc, "bank0", [this] (offs_t, u8& data, u8){ this->bank_w<0>(data); });
+	io_space().install_write_tap(0xfd, 0xfd, "bank1", [this] (offs_t, u8& data, u8){ this->bank_w<1>(data); });
+	io_space().install_write_tap(0xfe, 0xfe, "bank2", [this] (offs_t, u8& data, u8){ this->bank_w<2>(data); });
+	io_space().install_write_tap(0xff, 0xff, "bank3", [this] (offs_t, u8& data, u8){ this->bank_w<3>(data); });
+
+	// Not all memory mapper expansions allow reading of mapper registers.
+	io_space().install_read_tap(0xfc, 0xfc, "bank0", [this] (offs_t, u8& data, u8){ if (!machine().side_effects_disabled()) data = (~(m_nr_banks - 1)) | m_rambank[0]->entry(); });
+	io_space().install_read_tap(0xfd, 0xfd, "bank1", [this] (offs_t, u8& data, u8){ if (!machine().side_effects_disabled()) data = (~(m_nr_banks - 1)) | m_rambank[1]->entry(); });
+	io_space().install_read_tap(0xfe, 0xfe, "bank2", [this] (offs_t, u8& data, u8){ if (!machine().side_effects_disabled()) data = (~(m_nr_banks - 1)) | m_rambank[2]->entry(); });
+	io_space().install_read_tap(0xff, 0xff, "bank3", [this] (offs_t, u8& data, u8){ if (!machine().side_effects_disabled()) data = (~(m_nr_banks - 1)) | m_rambank[3]->entry(); });
+
+	page(0)->install_readwrite_bank(0x0000, 0x3fff, m_rambank[0]);
+	page(1)->install_readwrite_bank(0x4000, 0x7fff, m_rambank[1]);
+	page(2)->install_readwrite_bank(0x8000, 0xbfff, m_rambank[2]);
+	page(3)->install_readwrite_bank(0xc000, 0xffff, m_rambank[3]);
+
+	m_rambank[0]->set_entry(3);
+	m_rambank[1]->set_entry(2);
+	m_rambank[2]->set_entry(1);
+	m_rambank[3]->set_entry(0);
+}
+
+template <int Bank>
+void msx_cart_base_mm_ram_device::bank_w(u8 data)
+{
+	m_rambank[Bank]->set_entry(data % m_nr_banks);
+}
```

Comment:
As with the other cartridges, this should either be reduced to a mask or, if there are really potentially devices with non power-of-two sizes, reworked to deal with that.

Reply:
I have changed this to use map_non_power_of_two to configure the ram banks.